### ISSUE TITLE
Show more informative error when IdP denies user access

### ISSUE THREAD

**Issue by surfnet-niels** [OpenConext/OpenConext-engineblock#61]
In case an IdP denies a user access to a resource, this may be signaled using:

samlp:StatusCode Value="urn:oasis:names:tc:SAML:2.0:status:Responder"
           samlp:StatusCode Value="urn:oasis:names:tc:SAML:2.0:status:RequestDenied"

When this occurs, engine throws an error which is technically incorrect ("RequestDenied" is a valid SAML statement, not an error!) and is also not informative nor understandable by an enduser.

![error](https://cloud.githubusercontent.com/assets/1491826/3808789/0253153a-1c7c-11e4-84ea-96ace7b3a4bf.png)

In case urn:oasis:names:tc:SAML:2.0:status:RequestDenied is signaled, the enduser should be told to contact the IdP, as this is not a generic error, but a specific message that the user is denied access by the IdP

**Comment by thijskh** [comment]
Apparently this does not work. An authentication response with the following status code:

```xml
    <samlp:Status>
        <samlp:StatusCode Value="urn:oasis:names:tc:SAML:2.0:status:Responder">
            <samlp:StatusCode Value="urn:oasis:names:tc:SAML:2.0:status:RequestDenied" />
        </samlp:StatusCode>
    </samlp:Status>
```
generates "Error - An error occured".

**Comment by thijskh** [comment]
The condition is detected:
```
Oct 23 15:00:28 t03 EBLOG[8427]: [2015-10-23 15:00:28] engineblock.NOTICE: Received an Error Response {"request_id":"562a2f6c0f431","session_id":"amUdW4jrkwYmRH9wcZkosiVDAU7"} {"exception_message":"Responder: SimpleSAML_Error_Exception: Something is wrong...","status":"urn:oasis:names:tc:SAML:2.0:status:Responder","substatus":null,"status_message":"SimpleSAML_Error_Exception: Something is wrong..."}
```

However the code defers to the Corto input filter for actually handling it. While calling the filter, EB bombs out on missing an assertion:
```
Oct 23 15:00:28 t03 EBLOG[8427]: [2015-10-23 15:00:28] engineblock.ERROR: No assertions in response? {"request_id":"562a2f6c0f431","session_id":"amUdW4jrkwYmRH9wcZkosiVDAU7"} {"exception":{"severity":"ERROR","location":"/var/opt/openconext/OpenConext-engineblock-4.6.1/library/EngineBlock/Dispatcher.php:79","userId":null,"idp":null,"sp":null,"details":"EngineBlock_Exception: No assertions in response?\n#0 /var/opt/openconext/OpenConext-engineblock-4.6.1/library/EngineBlock/Dispatcher.php(48): EngineBlock_Dispatcher->_handleDispatchException(Object(RuntimeException))\n#1 /var/opt/openconext/OpenConext-engineblock-4.6.1/www/authentication/index.php(13): EngineBlock_Dispatcher->dispatch()\n#2 {main}\n\nRuntimeException: No assertions in response?\n#0 /var/opt/openconext/OpenConext-engineblock-4.6.1/library/EngineBlock/Corto/ProxyServer.php(796): EngineBlock_Saml2_ResponseAnnotationDecorator->getAssertion()\n#1 /var/opt/openconext/OpenConext-engineblock-4.6.1/library/EngineBlock/Corto/ProxyServer.php(772): EngineBlock_Corto_ProxyServer->callAttributeFilter(Array, Object(EngineBlock_Saml2_ResponseAnnotationDecorator), Object(EngineBlock_Saml2_AuthnRequestAnnotationDecorator), Object(OpenConext\\Component\\EngineBlockMetadata\\Entity\\ServiceProvider), Object(OpenConext\\Component\\EngineBlockMetadata\\Entity\\IdentityProvider))\n#2 /var/opt/openconext/OpenConext-engineblock-4.6.1/library/EngineBlock/Corto/Module/Service/AssertionConsumer.php(40): EngineBlock_Corto_ProxyServer->filterInputAssertionAttributes(Object(EngineBlock_Saml2_ResponseAnnotationDecorator), Object(EngineBlock_Saml2_AuthnRequestAnnotationDecorator))\n#3 /var/opt/openconext/OpenConext-engineblock-4.6.1/library/EngineBlock/Corto/Module/Services.php(56): EngineBlock_Corto_Module_Service_AssertionConsumer->serve('assertionConsum...')\n#4 /var/opt/openconext/OpenConext-engineblock-4.6.1/library/EngineBlock/Corto/ProxyServer.php(331): EngineBlock_Corto_Module_Services->serve('assertionConsum...')\n#5 /var/opt/openconext/OpenConext-engineblock-4.6.1/librar
```

**Comment by relaxnow** [comment]
Testing with https://github.com/OpenConext/OpenConext-engine-test-stand/blob/a67fdccc8aedbc2e4ea0dcc586583deb97cc5c4f/src/OpenConext/Bundle/FunctionalTestingBundle/Features/ClearErrorMessages.feature#L35-L48 on the latest OpenConext-engineblock#develop gives me:
```
|  <div class="box">
|    <div class="mod-content">
|      <h1>Error - Identity Provider error</h1>
|      <p>
|          Your Identity Provider sent an authentication response with an error status code.
|      </p>
|      <div class="l-overflow">
|        <table class="comp-table">
|          <thead></thead>
|          <tbody>
|                      <tr>
|                  <td class="strong">Status Code:</td>
|                  <td style="font-size: 0.8em;">Responder/RequestDenied</td>
|              </tr>
|                          <tr>
|                  <td class="strong">Status Message:</td>
|                  <td style="font-size: 0.8em;">Invalid IP range</td>
|              </tr>
|                          <tr>
|                  <td class="strong">Timestamp:</td>
|                  <td style="font-size: 0.8em;">2015-11-19T10:24:17+01:00</td>
|              </tr>
|                          <tr>
|                  <td class="strong">Unique Request ID:</td>
|                  <td style="font-size: 0.8em;">564d954120e48</td>
|              </tr>
|                          <tr>
|                  <td class="strong">User Agent:</td>
|                  <td style="font-size: 0.8em;">Symfony2 BrowserKit</td>
|              </tr>
|                          <tr>
|                  <td class="strong">IP Address:</td>
|                  <td style="font-size: 0.8em;">127.0.167.2</td>
|              </tr>
|                          <tr>
|                  <td class="strong">Service Provider:</td>
|                  <td style="font-size: 0.8em;">https://engine-test-stand.demo.openconext.org/Dummy%20SP/metadata</td>
|              </tr>
|                          <tr>
|                  <td class="strong">Identity Provider:</td>
|                  <td style="font-size: 0.8em;">https://engine-test-stand.demo.openconext.org/Dummy%20Idp/metadata</td>
|              </tr>
|                      </tbody>
|        </table>
|      </div>
```

Could you test again and maybe we could debug this together if it still fails for you?

**Comment by thijskh** [comment]
I can reproduce the problem selecting thkidp on https://profile.test.surfconext.nl. That IdP will always return the following:
```
    <samlp:Status>
        <samlp:StatusCode Value="urn:oasis:names:tc:SAML:2.0:status:Responder" />
        <samlp:StatusMessage>SimpleSAML_Error_Exception: Something is wrong...</samlp:StatusMessage>
    </samlp:Status>
```
and I do not get a specific error message.

The problem is also present in EB 4.5, and also with the SAML message content I quoted on Sep 11. That message can be reproduced with ADFS.

I can't really debug interactively right now or check the up to the minute state of develop but maybe this helps you to reprodue it.

**Comment by thijskh** [comment]
Two more real life examples of responses that generate "Error - An error occurred" instead of the intended error message.

```xml
<samlp:Response xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol" xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion" ID="_2741c719-f2ae-4689-a0c5-2e88a64d3bc1" Version="2.0" IssueInstant="2016-01-20T08:55:22Z" Destination="https://engine.surfconext.nl/authentication/sp/consume-assertion" InResponseTo="CORTO9f75fbf5c7ef4ec7531b63f62b18f77a04431ad6">
  <saml:Issuer>http://federation.voorbeeld.nl/adfs/services/trust</saml:Issuer>
  <samlp:Status>
    <samlp:StatusCode Value="urn:oasis:names:tc:SAML:2.0:status:Responder">
      <samlp:StatusCode Value="urn:oasis:names:tc:SAML:2.0:status:RequestDenied"/>
    </samlp:StatusCode>
  </samlp:Status>
</samlp:Response>
```

```xml
<samlp:Response xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol" xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion" ID="id-HW1q4BUkjB3GuVvTlqYtwr1cuuI-" Version="2.0" IssueInstant="2016-01-20T20:28:02Z" Destination="https://engine.surfconext.nl/authentication/sp/consume-assertion" InResponseTo="CORTO8a275030e97351e68e7cc0f89d5b46393d9ee3d9">
  <saml:Issuer>https://voorbeeld.nl/</saml:Issuer>
  <samlp:Status>
    <samlp:StatusCode Value="urn:oasis:names:tc:SAML:2.0:status:Requester">
      <samlp:StatusCode Value="urn:oasis:names:tc:SAML:2.0:status:RequestDenied"/>
    </samlp:StatusCode>
    <samlp:StatusMessage>The AuthnRequest could not be validated</samlp:StatusMessage>
  </samlp:Status>
</samlp:Response>
```

**Comment by thijskh** [comment]
Engine Test Stand functional tests pass because they include an `<Assertion>...</Assertion>` block in the response used for testing. Status Responses (not having Success) do not include such a block normally. Therefore in Engine Test Stand, the premature Exception "No assertions in response?" is not triggered.

**Comment by thijskh** [comment]
The behaviour has changed somewhat between 5.0 and 5.1, and the error is now "Invalid Identity Provider response". It's still not the desired behaviour though. That is to show the screen like in mentioned in https://github.com/OpenConext/OpenConext-engineblock/issues/61#issuecomment-60252915 which includes the error code and error text as supplied by the IdP.

**Comment by arothuis** [comment]
This has been fixed in #355.
